### DISCUSSION TITLE
RespondJSONのbodyに指定しているErrResponseと&ErrResponseの使い分けについて

### DISCUSSION THREAD

**Discussion by youta32449999** [budougumi0617/go_todo_app#60] [Q&A] [answered]
エラーレスポンスを返す際に`RespondJSON`の`body`を`ErrResponse`としているものと`&ErrResponse`としているものがあるのですが、こちらはどのような使い分けとなっているのでしょうか?

大変恐縮ですがご教授いただけますと幸いです。どうぞよろしくお願いいたします。
```go
// /handler/login.go
RespondJSON(ctx, w, &ErrResponse{
	Message: err.Error(),
}, http.StatusInternalServerError)

// /handler/middleware.go
RespondJSON(r.Context(), w, ErrResponse{
	Message: "not find auth info",
	Details: []string{err.Error()},
}, http.StatusUnauthorized)
```

**Comment by budougumi0617** [comment, marked as answer]
一般論としては参照渡し（`&ErrResponse`）を使っていたほうが無難かなと思います。

1. メモリアロケーションが発生しない
    - 大きい構造体を渡したときのことを考慮しなくて済む。
2. 呼び出した関数内で「更新したつもりで更新してなかった」が防げる
    - これは賛否両論あるかと思いますが…

`XxxMiddleware` 関数の実装内でコピー渡し（`ErrResponse`）をしているのは渡す`ErrResponse`型の値のサイズが明らかだから、上記1の観点でコピー渡ししても大したメモリアロケーションにならない」と判断してコピー渡しにしていたかと思います。
その他の実装では `ErrResponse` のサイズが内部の実装によって大きくなる可能性があるので `&ErrResponse` にしていました。

「 `RespondJSON` の中で `ErrResponse` を更新することはないのでコピー渡しのほうがよいのでは」という見方もあるかもしれませんし、「一律 `&ErrResponse` に合わせておいたほうが使い分け悩まずにすむんじゃない」という見方もできますね。
あまり回答になっていないかもしれませんが、私が考えていた意図は上記のような感じです…!!

**Reply by youta32449999** [reply]
>メモリアロケーションが発生しない
大きい構造体を渡したときのことを考慮しなくて済む。

>XxxMiddleware 関数の実装内でコピー渡し（ErrResponse）をしているのは渡すErrResponse型の値のサイズが明らかだから、上記1の観点でコピー渡ししても大したメモリアロケーションにならない」と判断してコピー渡しにしていたかと思います。
その他の実装では ErrResponse のサイズが内部の実装によって大きくなる可能性があるので &ErrResponse にしていました。

こちらの内容とても参考になりました!まさに「更新することはないのでコピー渡しのほうがよいのでは」のように考えておりました。大きな構造体を渡しても大きなメモリの確保が行われないようにするためなのですね!

丁寧なご回答ありがとうございます!
